File: implementations/go/provekit-ir-symbolic/verifier/runner.go
package verifier

// Runner is the orchestrator. It composes the 6 stages into a single
// RunBridgeEnforcement entry point + uses goroutines + a channel to
// process per-callsite verifications in parallel (each callsite's
// resolve→instantiate→solve chain is independent, so they fan out).
type Runner struct {
	loadStage        *LoadAllProofsStage
	enumerateStage   *EnumerateCallsitesStage
	resolveStage     *ResolveTargetStage
	instantiateStage *InstantiateStage
	solveStage       *SolveObligationStage
	reportStage      *ReportStage
	solver           *Solver
}

// NewRunner constructs a Runner with the configured solver.
func NewRunner(solver *Solver) *Runner {
	return &Runner{
		loadStage:        &LoadAllProofsStage{},
		enumerateStage:   &EnumerateCallsitesStage{},
		resolveStage:     &ResolveTargetStage{},
		instantiateStage: &InstantiateStage{},
		solveStage:       NewSolveObligationStage(),
		reportStage:      &ReportStage{},
		solver:           solver,
	}
}

// RunBridgeEnforcement executes the full pipeline:
//  1. load-all-proofs
//  2. enumerate-callsites
//     3+4+5. for each callsite (in parallel): resolve, instantiate, solve
//  6. report
func (r *Runner) RunBridgeEnforcement(projectRoot string) (*Report, error) {
	pool, err := r.loadStage.Run(projectRoot)
	if err != nil {
		return nil, err
	}
	callsites := r.enumerateStage.Run(pool)

	// Fan out per-callsite work via goroutines + a channel. Order is
	// preserved by indexing into a result slice.
	rows := make([]ReportRow, len(callsites))
	type completion struct {
		idx int
		row ReportRow
	}
	ch := make(chan completion, len(callsites))
	for i, cs := range callsites {
		go func(idx int, cs CallSite) {
			ch <- completion{idx: idx, row: r.processCallSite(cs, pool)}
		}(i, cs)
	}
	for i := 0; i < len(callsites); i++ {
		c := <-ch
		rows[c.idx] = c.row
	}
	return r.reportStage.Run(rows, pool.LoadErrors), nil
}

// processCallSite runs a single callsite's resolve→instantiate→solve
// chain. Returns one ReportRow.
func (r *Runner) processCallSite(cs CallSite, pool *MementoPool) ReportRow {
	resolved, fail := r.resolveStage.Run(cs, pool)
	if resolved == nil {
		return ReportRow{CallSite: cs, Status: "unresolved-target", Reason: fail}
	}
	if cs.ArgTerm == nil {
		return ReportRow{CallSite: cs, Status: "unliftable-argument", Reason: "no first arg"}
	}
	obligation, err := r.instantiateStage.Run(resolved.IRFormula, cs.ArgTerm)
	if err != nil {
		return ReportRow{CallSite: cs, Status: "lift-error", Reason: err.Error()}
	}

	// Tier 0: Check if obligation formula itself is verified in the pool.
	if _, ok := pool.Verify(obligation); ok {
		return ReportRow{CallSite: cs, Status: "discharged", Reason: "tier0: memento-is-verification"}
	}

	// Tier 0c: Check if post → pre implication is already proven.
	// This requires computing CIDs for the post and pre formulas.
	// For now, we skip to Tier 3 (solver) if no direct verification.
	// Future: extract postHash from bridge and preHash from resolved.

	solveResult, err := r.solveStage.Run(obligation, r.solver)
	if err != nil {
		return ReportRow{CallSite: cs, Status: "lift-error", Reason: err.Error()}
	}
	return ReportRow{CallSite: cs, Status: string(solveResult.Verdict)}
}

File: implementations/cpp/provekit/verifier/runner.cpp
// SPDX-License-Identifier: Apache-2.0

#include "runner.hpp"

#include "load_all_proofs.hpp"
#include "enumerate_callsites.hpp"
#include "resolve_target.hpp"
#include "instantiate.hpp"
#include "smt_emitter.hpp"
#include "solve_obligation.hpp"
#include "report.hpp"

#include <future>
#include <vector>

namespace provekit::verifier {

namespace {

struct PerCallSite {
    CallSite cs;
    ObligationVerdict verdict = ObligationVerdict::Undecidable;
    std::string reason;
};

PerCallSite work_one(const CallSite& cs,
                      const MementoPool& pool,
                      const std::string& z3_path) {
    PerCallSite r;
    r.cs = cs;

    ResolveTargetStage rs;
    auto rr = rs.Run(cs, pool);
    if (!rr.ok) {
        r.verdict = ObligationVerdict::Undecidable;
        r.reason = "resolve-target: " + rr.error;
        return r;
    }

    InstantiateStage is;
    Obligation ob;
    std::string err;
    if (!is.Run(rr.resolved, cs.arg_term, &ob, &err)) {
        r.verdict = ObligationVerdict::Undecidable;
        r.reason = "instantiate: " + err;
        return r;
    }

    SmtEmitter se;
    err.clear();
    std::string smt = se.Emit(ob.ir_formula, &err);
    if (!err.empty()) {
        r.verdict = ObligationVerdict::Undecidable;
        r.reason = "smt-emit: " + err;
        return r;
    }

    SolveObligationStage solver(z3_path);
    auto sr = solver.Run(smt);
    r.verdict = sr.verdict;
    if (!sr.error.empty()) {
        r.reason = sr.error;
    } else if (sr.verdict == ObligationVerdict::Unsatisfied) {
        r.reason = "solver returned sat (counterexample found): obligation falsifiable";
    } else if (sr.verdict == ObligationVerdict::Discharged) {
        r.reason = "solver returned unsat: obligation holds";
    }
    return r;
}

}  // namespace

Report Runner::Run() {
    Report report;

    LoadAllProofsStage stage1;
    MementoPool pool = stage1.Run(cfg_.project_root);

    EnumerateCallsitesStage stage2;
    std::vector<CallSite> callsites = stage2.Run(pool);

    // Fan stages 3-5 out per callsite.
    std::vector<std::future<PerCallSite>> futures;
    futures.reserve(callsites.size());
    const std::string z3 = cfg_.z3_path;
    for (const auto& cs : callsites) {
        futures.push_back(std::async(std::launch::async, [&pool, cs, z3]() {
            return work_one(cs, pool, z3);
        }));
    }

    ReportStage stage6;
    for (auto& f : futures) {
        PerCallSite r = f.get();
        stage6.Add(r.cs, r.verdict, r.reason, &report);
    }
    stage6.AddLoadErrors(pool.load_errors, &report);

    return report;
}

}  // namespace provekit::verifier
